### PR TITLE
Block explorer links

### LINKED ISSUE #899: Add StarkScan as block explorer when deploying
As a StarkNet smart contract developer, I want to have a link to the StarkScan ([Twitter](https://twitter.com/StarkscanCo/status/1575170281416454154)) block explorer ([testnet](https://testnet.starkscan.co/), [mainnet](https://starkscan.co/)) when deploying a contract so that I can choose to use either voyager or starkscan.

The photo below shows how the voyager link is given when deploying a contract. Would be nice to have a link to starkscan too. 

image

### PR DESCRIPTION
**Pull request by kasperski95**
Introduce Block Explorer classes. Integrate with existing commands.
Closes #899

### CODE REVIEW

File: protostar/cli/protostar_arg_type.py [hunk 1]
```diff
@@ -52,3 +55,8 @@ def parse_hex_or_decimal(arg: str) -> int:
     if arg.startswith("0x"):
         return int(arg, 16)
     return int(arg)
+
+
+def parse_block_explorer_type(arg: str) -> SupportedBlockExplorerName:
+    assert arg in SUPPORTED_BLOCK_EXPLORER_NAMES
```
Comment (mkaput) [inline protostar/cli/protostar_arg_type.py:61]: could we raise ValueError here?

File: protostar/cli/protostar_arg_type_test.py [hunk 2]
```diff
@@ -39,3 +39,15 @@ def test_address_arg_type():
...
+
+def test_block_explorer_type():
+    parser = map_protostar_type_name_to_parser("block_explorer")
+    result = parser("starkscan")
+    assert result == "starkscan"
+
+
+def test_fail_on_wrong_block_explorer():
+    parser = map_protostar_type_name_to_parser("block_explorer")
+    with pytest.raises(AssertionError):
```
Comment (mkaput) [inline protostar/cli/protostar_arg_type_test.py:52]: pls match error message here

File: protostar/starknet_gateway/block_explorer/fake_block_explorer.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+class FakeBlockExplorer(BlockExplorer):
+    def create_link_to_transaction(self, tx_hash: TransactionHash) -> URL:
+        return ""
+
+    def create_link_to_contract(self, contract_address: ContractAddress) -> URL:
+        return ""
+
+    def create_link_to_class(self, class_hash: ClassHash) -> URL:
+        return ""
```
Comment (mkaput) [inline protostar/starknet_gateway/block_explorer/fake_block_explorer.py:18]: I think we can avoid this class?

Why not assume voyager as default block explorer even in tests? Simply avoid testing block explorer output in assertions.

File: protostar/starknet_gateway/block_explorer/block_explorer_factory.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+SUPPORTED_BLOCK_EXPLORER_NAMES: list[SupportedBlockExplorerName] = [
+    "starkscan",
+    "viewblock",
+    "voyager",
+]
+
+
+def create_block_explorer(
+    block_explorer_name: SupportedBlockExplorerName,
+    network: Optional[PredefinedNetwork],
```
Comment (mkaput) [inline protostar/starknet_gateway/block_explorer/block_explorer_factory.py:20]: I think `network` could be just an argument to all `create_*` functions and thus you could make `BlockExplorer` instances parameterless.

File: protostar/starknet_gateway/block_explorer/stark_scan_block_explorer_test.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
+import pytest
```
Comment (mkaput) [inline protostar/starknet_gateway/block_explorer/stark_scan_block_explorer_test.py:1]: I think this test does not bring any value. Tested functions are trivial by definition? Are we sure we want this?

File: protostar/starknet_gateway/block_explorer/stark_scan_block_explorer.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    TransactionHash,
+)
+
+
+class StarkScanBlockExplorer(BlockExplorer):
+    def __init__(self, network: PredefinedNetwork) -> None:
+        super().__init__()
+        network_to_domain: dict[PredefinedNetwork, URL] = {
+            "mainnet": "https://starkscan.co",
+            "testnet": "https://testnet.starkscan.co",
```
Comment (mkaput) [inline protostar/starknet_gateway/block_explorer/stark_scan_block_explorer.py:17]: What about testnet2?

**Review by mkaput — CHANGES_REQUESTED** [review]

File: protostar/starknet_gateway/block_explorer/fake_block_explorer.py (hunk 3 above)
Comment (mkaput) [reply protostar/starknet_gateway/block_explorer/fake_block_explorer.py:18]: discussed this privately

File: website/docs/cli-reference.md [hunk 7]
```diff
@@ -53,6 +53,11 @@ Required.
 Path to compiled contract.
 #### `--account-address ADDRESS`
 Account address.
+#### `--block-explorer BLOCK_EXPLORER=starkscan`
+Generated links will point to that block explorer. Available values:
+- starkscan
+- viewblock
+- voyager
```
Comment (Arcticae) [inline website/docs/cli-reference.md:60]: I think users should be somehow aware that we can add one on their request (after verification, if the project is mature enough). We should either include it here or somewhere in the docs

**Review by Arcticae — APPROVED** [review]
Looks good codewise. I am wondering if we should either hurry with this or wait for @mkaput to finish his formatter output thing, but either way if that isn't conflicting - i think it's mergable as-is.

**Comment by mkaput** [comment]
I would prefer to have this merged before I get with formatter stuff onto it

File: protostar/starknet_gateway/block_explorer/stark_scan_block_explorer.py (hunk 6 above)
Comment (kasperski95) [reply protostar/starknet_gateway/block_explorer/stark_scan_block_explorer.py:17]: It's out of scope of this PR. Starknet.py needs to introduce it.

File: website/docs/cli-reference.md (hunk 7 above)
Comment (kasperski95) [reply website/docs/cli-reference.md:60]: This issue was requested by users. I am not sure if it's needed.

File: protostar/starknet_gateway/block_explorer/stark_scan_block_explorer.py (hunk 6 above)
Comment (mkaput) [reply protostar/starknet_gateway/block_explorer/stark_scan_block_explorer.py:17]: Starknet.py team says they will rather not introduce anything to PredefinedNetwork. But I think this doesn't mean we should also not?

File: protostar/starknet_gateway/block_explorer/block_explorer_factory.py (hunk 4 above)
Comment (kasperski95) [reply protostar/starknet_gateway/block_explorer/block_explorer_factory.py:20]: I'd have to then pass network to formatters.

File: protostar/cli/protostar_arg_type_test.py (hunk 2 above)
Comment (kasperski95) [reply protostar/cli/protostar_arg_type_test.py]: Changed to `ValueError`. argparse generates an error message for ValueErrors.

```
... argument --block-explorer: invalid parse_block_explorer_type value: 'abc'
```

File: protostar/starknet_gateway/block_explorer/stark_scan_block_explorer_test.py (hunk 5 above)
Comment (kasperski95) [reply protostar/starknet_gateway/block_explorer/stark_scan_block_explorer_test.py]: Removed tests

File: protostar/starknet_gateway/block_explorer/block_explorer_factory.py (hunk 4 above)
Comment (mkaput) [reply protostar/starknet_gateway/block_explorer/block_explorer_factory.py:20]: OK, makes sense

File: protostar/starknet_gateway/block_explorer/stark_scan_block_explorer.py (hunk 6 above)
Comment (kasperski95) [reply protostar/starknet_gateway/block_explorer/stark_scan_block_explorer.py:19]: Then we need to decouple this concept from Starknet.py. That's a different issue.

**Pull request merged**